File: backend/app/schemas/cover_letter.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class CoverLetterGenerateRequest(BaseModel):
    job_id: uuid.UUID | None = None
    job_description: str | None = Field(default=None, max_length=10000)
    tone: str | None = Field(default="professional", pattern="^(professional|casual|enthusiastic)$")

    @model_validator(mode="after")
    def check_at_least_one_source(self) -> "CoverLetterGenerateRequest":
        if self.job_id is None and not self.job_description:
            raise ValueError("At least one of job_id or job_description must be provided")
        return self


class CoverLetterGenerateManualRequest(BaseModel):
    job_description: str = Field(min_length=50, max_length=10000)
    company_name: str | None = Field(default=None, max_length=200)
    job_title: str | None = Field(default=None, max_length=200)
    tone: str | None = Field(default="professional", pattern="^(professional|casual|enthusiastic)$")


class CoverLetterGenerateResponse(BaseModel):
    cover_letter_id: uuid.UUID
    status: str
    content_json: dict | None = None


class CoverLetterStatusResponse(BaseModel):
    cover_letter_id: uuid.UUID
    status: str
    tone: str | None = None


class CoverLetterListItem(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID | None = None
    job_title: str | None = None
    company_name: str | None = None
    tone: str | None = None
    created_at: datetime


class CoverLetterListResponse(BaseModel):
    cover_letters: list[CoverLetterListItem]
    total: int


class CoverLetterResponse(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID | None = None
    job_title: str | None = None
    company_name: str | None = None
    content: str | None = None
    content_json: dict | None = None
    tone: str | None = None
    ai_model_used: str | None = None
    status: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

File: frontend/src/types/cover-letter.ts
export interface CoverLetterGenerateRequest {
  job_id?: string;
  job_description?: string;
  tone?: "professional" | "casual" | "enthusiastic";
  force_regenerate?: boolean;
}

export interface CoverLetterGenerateManualRequest {
  job_description: string;
  company_name?: string;
  job_title?: string;
  tone?: "professional" | "casual" | "enthusiastic";
}

export interface CoverLetterGenerateResponse {
  cover_letter_id: string;
  status: "generating" | "completed" | "failed";
  content_json: Record<string, unknown> | null;
}

export interface CoverLetterStatusResponse {
  cover_letter_id: string;
  status: "generating" | "completed" | "failed";
  tone: string | null;
}

export interface CoverLetterListItem {
  id: string;
  job_id: string | null;
  job_title: string | null;
  company_name: string | null;
  tone: string | null;
  created_at: string;
}

export interface CoverLetterListResponse {
  cover_letters: CoverLetterListItem[];
  total: number;
}

export interface CoverLetterResponse {
  id: string;
  job_id: string | null;
  job_title: string | null;
  company_name: string | null;
  content: string | null;
  content_json: Record<string, unknown> | null;
  pdf_url?: string | null;
  tone: string | null;
  ai_model_used: string | null;
  status: string | null;
  created_at: string;
  updated_at: string | null;
}

export interface CoverLetterParagraph {
  heading: string | null;
  body: string;
}

export interface CoverLetterContent {
  paragraphs: CoverLetterParagraph[];
  tone_used: string;
  keywords_addressed: string[];
  full_text: string;
}
